lib/system-nixpkgs-map.nix

```diff
@@ -78,6 +78,7 @@ pkgs:
      vulkan = pkgs.vulkan-loader;
      sodium = pkgs.libsodium;
      gfortran = pkgs.gfortran.cc.lib;
+     blas = pkgs.openblasCompat;
```

Comment:
I just realized that while this solves the issue with missing header file it creates another issue with missing symbol at the time executable gets link so it's not a full fix to the issue.

Reply:
does openblasCompat come with a companion package that contains the symbols? Or does the openblasCompat package need to be customized to include that symbol?

Reply:
It does include both headers and shared object. This is what I used to check this:

```bash
$ nix-shell -p openblasCompat pkgconfig --pure

[nix-shell:~]$ pkg-config --cflags openblas
-I/nix/store/vggipyxqc69inb2fk4wchfd77ahlh4zr-openblas-0.3.10/include

[nix-shell:~]$ ls /nix/store/vggipyxqc69inb2fk4wchfd77ahlh4zr-openblas-0.3.10/include
cblas.h  f77blas.h  lapack.h  lapacke.h  lapacke_config.h  lapacke_mangling.h  lapacke_utils.h  openblas_config.h
  

[nix-shell:~]$ pkg-config --libs openblas
-L/nix/store/vggipyxqc69inb2fk4wchfd77ahlh4zr-openblas-0.3.10/lib -lopenblas

[nix-shell:~]$ ls /nix/store/vggipyxqc69inb2fk4wchfd77ahlh4zr-openblas-0.3.10/lib
cmake       libblas.so.3  libcblas.so.3  liblapack.so.3  liblapacke.so.3  libopenblas.so.0         pkgconfig
libblas.so  libcblas.so   liblapack.so   liblapacke.so   libopenblas.so   libopenblasp-r0.3.10.so
```

I think the problem is with the package `hmatrix-morpheus` itself. It doesn't use pkgconfig but uses dumb/naive way of linking to native dependencies:

https://github.com/aligusnet/morpheus/blob/master/hmatrix-morpheus.cabal#L65-L73

My current work-around is to specify is using `depends` on package module (minimal example):

```nix
haskell-nix.project {
  src = ./.;
  modules = [
    { packages.hmatrix-morpheus.components.library.configureFlags = [
        "-f openblas"
        "--extra-lib-dirs=${openblas}/lib"
      ];
      packages.hmatrix-morpheus.components.library.depends = [ openblas ];
    }
  ];
  projectFileName = "stack.yaml";
}
```

not that I'm **NOT** using the overlay to map `blas = openblasCompat`. Instead I use `--extra-lib-dirs` to make header file available. The `depends` is then a hack to make object available - without it there is an error with missing symbols when executable is getting linked.

After spending quite some time on this I have a 2 questions:

1. What is the correct way of doing these types of overrides with haskell.nix?
2. What is the correct way to path this in upstream so my fix can be up streamed?

Reply:
So haskell.nix in general assuems that the cabal file is the ground truth, as such all info is picked from the cabal file. [extra-libraries: openblas](https://cabal.readthedocs.io/en/3.4/cabal-package.html#pkg-field-extra-libraries), is actually the correct way. However you need to change your flag definition, such that haskell.nix can see it.
```
modules = [
  packages.hmatrix-morpheus.flags.openblas = true;
];
```
that should bring in the dependencies properly. the derivations then depend on openblas, and nix will put the header into the include path, as well as the libraries into the library search path.

Reply:
Thanks for help. I just tried this and it works :+1: